File: C++/046. Permutations.cpp
#include <iostream>
#include <algorithm>
#include <vector>

class Solution {
public:
    std::vector<std::vector<int>> permute(std::vector<int>& nums) {
        sort(nums.begin(), nums.end());
        std::vector<std::vector<int>> res;
        while (true) {
            res.emplace_back(nums);
            int i, j, tmp;
            for (i = int(nums.size()) - 2; i >= 0 && nums[i] > nums[i + 1]; i--);
            if (i < 0)
                break;
            for (j = i + 1; j < int(nums.size()) && nums[j] > nums[i]; j++);
            tmp = nums[i];
            nums[i] = nums[j - 1];
            nums[j - 1] = tmp;
            for (i = i + 1, j = int(nums.size()) - 1; i < j; i++, j--) {
                tmp = nums[i];
                nums[i] = nums[j];
                nums[j] = tmp;
            }
        }
        return res;
    }
};

File: Go/046. Permutations.go
package main

import "sort"

func permute(nums []int) [][]int {
	sort.Ints(nums)
	res := make([][]int, 0, 0)
	for true {
		tmp := make([]int, len(nums), len(nums))
		copy(tmp, nums)
		res = append(res, tmp)
		var i, j int
		for i = len(nums) - 2; i >= 0 && nums[i] > nums[i + 1]; i-- {}
		if i < 0 {
			break
		}
		for j = i + 1; j < len(nums) && nums[j] > nums[i]; j++ {}
		nums[i], nums[j - 1] = nums[j - 1], nums[i]
		for i, j = i + 1, len(nums) - 1; i < j; i, j = i + 1, j - 1 {
			nums[i], nums[j] = nums[j], nums[i]
		}
	}
	return res
}